### PR TITLE
Moonpile/issue89

### PR DESCRIPTION
**Pull request by Moonpile**
- Made "Improve" context menu on skills always appear
- Dialog with up to four options for improvement

### CODE REVIEW

**Review by wake42 — CHANGES_REQUESTED** [review]
Hi!
I have some comments. First I don't think we should give page references to the core book since the user might not have that one - they might be playing with the starter-set or even the quick start rules.

Second I think it should be radioboxes instead of checkboxes since checkboxes imply that you can select many while radioboxes imply that only one is selectable.

Finally I think the chance calculation is wrong. If I understand correctly you add the skill category bonus to the chance. In the rulebook they add that bonus to the roll instead. I was thinking about that though... What do you think about changing it so that you have to roll under 100 - learned chance instead (learned chance is the same thing as chance - skill category bonus on the skill item). It seems more logical to succeed a roll to improve instead of failing it, even if it's not exactly according to how the rulebook puts it.

**Comment by Moonpile** [comment]
I think the page reference was removed before I even made the Pull Request.  Perhaps it's in one of the incremental commits?

The experience roll calculation is 1d100+ categoryMod, and if this is greater than the skill chance (total) or greater than or equal to 100, then you gain.

A few things I think we may want to think through better in the future:

- if you add "content" to a roll it "overwrites" the actual dice result.  Therefore on rolls if we have more to say than just a little bit we need to put it in the "flavor" with some HTML around it to make it look nice.  We should come up with a standard for this, or perhaps even an HBS template.
- I put some "style" elements on the divs in the dialog.  If we do more dialogs perhaps we can extract our look and feel to scss.

**Review by wake42 — COMMENTED** [review]
I think it's more interesting to show the chance of success than the actual mechanics behind it. What do you think about this sketch with less text. Also with radio buttons instead of checkboxes.
image

**Comment by wake42** [comment]
> * if you add "content" to a roll it "overwrites" the actual dice result.  Therefore on rolls if we have more to say than just a little bit we need to put it in the "flavor" with some HTML around it to make it look nice.  We should come up with a standard for this, or perhaps even an HBS template.

Yes the styling and general handling of chat messages as a whole would benefit from a work trough!

**Comment by Moonpile** [comment]
I had to put the chance to gain on another line or the experience h2 element wrapped and the training one didn't and that looked bad.

image
image
image
image

File: src/items/skill-item/skillSheet.ts [hunk 1]
```diff
@@ -108,4 +111,179 @@ export class SkillSheet extends RqgItemSheet<ItemSheet.Options, SkillSheetData |
...
+    var chanceToGain = 0;
+    if (chance > 100) {
+      chanceToGain = categoryMod;
+    } else {
+      chanceToGain = (100 - chance) + categoryMod;
+    }
+
+    const btnImprove = getGame().i18n.format("DIALOG.improveSkillDialog.btnDoImprovement");
+    const btnCancel = getGame().i18n.format("DIALOG.improveSkillDialog.btnCancel");
+    var buttons = {};
```
Comment (wake42) [inline src/items/skill-item/skillSheet.ts:140]: Just some syntax, but it you type this as `const buttons: any = {}` then you don't need the @ts-ignores further down.

Also generally, please don't use `var`. Use `const` wherever possible and `let` if that doesn't compile to keep variable scopes to a minimum :-)

File: src/items/skill-item/skillSheet.ts [hunk 2]
```diff
@@ -108,4 +111,179 @@ export class SkillSheet extends RqgItemSheet<ItemSheet.Options, SkillSheetData |
...
+          await SkillSheet.submitImproveSkillDialog(
+            html as JQuery,
+            actor,
+            item,
+            speakerName
+          )
+      }
+    }
+
+    //TODO: Not sure why this wouldn't work with "await"
```
Comment (wake42) [inline src/items/skill-item/skillSheet.ts:164]: The reason is that the `showImproveSkillDialog` function is not declared as async. Just add that and you can use await.

**Review by wake42 — CHANGES_REQUESTED** [review]
Looking good!
I found that the order of cancel / improve buttons was flipped. Foundry standard is having cancel to the right.
Also I think there is a UX problem if you don't select any radiobutton. Then it just clears the experience without any feedback. I think either disable the submit button until something is chosen, or preselect one option to begin with.
And finally the first "What type of improvement..." line is too long for the row. I would either reformulate to make it shorter, or make the dialog wider.

File: src/items/skill-item/skillSheet.ts (hunk 1 above)
Comment (Moonpile) [reply src/items/skill-item/skillSheet.ts:140]: The limits of my javascript and typescript knowledge are showing!

**Comment by Moonpile** [comment]
Ok, I think I've addressed everything including applying Prettier.

**Review by wake42 — APPROVED** [review]
Looking good!

**Pull request merged**